### PR TITLE
feature: adding distributed dbscan support without allgatherv(memory saving implementation)

### PR DESCRIPTION
**Pull request by Alexandr-Solovev**
# Description
Dbscan algorithm has been reimplemented with adding memory save clustering method.

### CODE REVIEW

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp [hunk 1]
```diff
@@ -41,74 +41,141 @@ static result_t compute_kernel_dense_impl(const context_gpu& ctx,
...
+    //array storages the information about neighbours of the point for distributed mode
     auto [arr_responses, responses_event] =
-        pr::ndarray<std::int32_t, 1>::full(queue, block_size, -1, sycl::usm::alloc::device);
-    auto [arr_queue, queue_event] =
-        pr::ndarray<std::int32_t, 1>::full(queue, row_count, -1, sycl::usm::alloc::device);
-    auto [arr_queue_front, queue_front_event] =
+        pr::ndarray<std::int32_t, 1>::full(queue, local_row_count, -1, sycl::usm::alloc::device);
+
+    //array storages the information about which point are core neighbours in the current step
+    auto [observation_indicies, observation_indicies_event] =
```
Comment (olegkkruglov) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp:101]:
```suggestion
    auto [observation_indices, observation_indices_event] =
```

**Review by Vika-F — COMMENTED** [review]
Thank you for making distributed DBSCAN implementation less memory-hungry.
I have 2 main concerns:

1. The amount of comments in SYCL kernels code makes it hard to understand the implementation.
2. It seems there is no test that runs distributed DBSCAN on sufficiently big dataset to check that this implementation actually consumes less memory than the previous one.

File: cpp/oneapi/dal/algo/dbscan/test/batch.cpp [hunk 2]
```diff
@@ -143,14 +137,14 @@ TEMPLATE_LIST_TEST_M(dbscan_batch_test,
     SKIP_IF(this->not_float64_friendly());
     using float_t = std::tuple_element_t<0, TestType>;
 
-    constexpr float_t data[] = { 0.0, 2.0, 3.0, 4.0, 6.0, 8.0, 10.0 };
-    const auto x = homogen_table::wrap(data, 7, 1);
+    constexpr float_t data[] = { 0.0, 2.0, 3.0, 4.0, 6.0, 8.0, 2.0, 2.0 };
+    const auto x = homogen_table::wrap(data, 8, 1);
```
[the remark this answers is not in the document]
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/algo/dbscan/test/batch.cpp:141]: It makes the test case little bit more difficult

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp [hunk 3]
```diff
@@ -41,139 +44,270 @@ static result_t compute_kernel_dense_impl(const context_gpu& ctx,
...
+        weights_nd_replace = pr::ndarray<Float, 2>::empty(queue,
+                                                          { max_local_block_size, 1 },
+                                                          sycl::usm::alloc::device);
+
+        copy(queue, weights_nd_replace, weights_nd, {}).wait_and_throw();
+    }
+    const Float epsilon = desc.get_epsilon() * desc.get_epsilon();
     const std::int64_t min_observations = desc.get_min_observations();
 
+    //array indicates is the point core or no
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp:111]: Most of the comments in oneDAL begin with a space. Please align the comments in this PR with the rest of the codebase.
 
```suggestion
    // array indicates if the point is a core point or not
```

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp [hunk 4]
```diff
@@ -41,139 +44,270 @@ static result_t compute_kernel_dense_impl(const context_gpu& ctx,
...
     auto [arr_cores, cores_event] =
-        pr::ndarray<std::int32_t, 1>::full(queue, block_size, 0, sycl::usm::alloc::device);
+        pr::ndarray<std::int32_t, 1>::full(queue, local_row_count, 0, sycl::usm::alloc::device);
+
+    //array storages the information about neighbours of the point for distributed mode
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/compute_kernel_brute_force_dpc.cpp:115]: "storages" -> "stores". The same applies for the rest of the comments in this file.
```suggestion
    // array stores the information about neighbours of the point in distributed mode
```

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp [hunk 5]
```diff
@@ -39,59 +39,87 @@ inline std::int64_t get_recommended_check_block_size(const sycl::queue& queue,
...
+///
+/// @param[in]  queue The SYCL queue
+/// @param[in]  data  The input data of size `row_count` x `column_count`
+/// @param[in]  weights  The input weights of size `row_count` x `1`
+/// @param[in]  cores  The current cores of size `row_count` x `1`
+/// @param[in]  neighbours  The current neighbours of size `row_count` x `1`
+/// it contains the counter of neighbours for each point
+/// @param[in]  epsilon  The input parameter epsilon
+/// @param[in]  min_observations  The input parameter min_observation
+/// @param[in]  deps  Events indicating availability of the `data` for reading or writing
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp:63]: Please align the descriptions of the parameters as it is done in the rest of the codebase.
The goal is to make the code more unified. It should look like everything was written by a single person.

```suggestion
/// @tparam Float        Floating-point type used to perform computations
/// @tparam use_weights  Bool type used to check that weights are enabled
///
/// @param[in]  queue            The SYCL queue
/// @param[in]  data             The input data of size `row_count` x `column_count`
/// @param[in]  weights          The input weights of size `row_count` x `1`
/// @param[in]  cores            The current cores of size `row_count` x `1`
/// @param[in]  neighbours       The current neighbours of size `row_count` x `1`
///                              it contains the counter of neighbours for each point
/// @param[in]  epsilon          The input parameter epsilon
/// @param[in]  min_observations The input parameter min_observation
/// @param[in]  deps             Events indicating availability of the `data` for reading or writing
```

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp [hunk 6]
```diff
@@ -107,167 +135,507 @@ struct get_core_wide_kernel {
...
                     }
                 });
         });
         return event;
     }
-    static constexpr std::int64_t min_width = 4;
 };
 
+///  A struct that finds the core points without subgroups
+///  it is effective only on narrow cases
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp:156]: Please specify what is meant here by a narrow case. Small number of columns?

File: cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp [hunk 7]
```diff
@@ -303,104 +672,194 @@ std::int32_t kernels_fp<Float>::start_next_cluster(sycl::queue& queue,
...
-                             queue_ptr[queue_front_ptr[0]] = start_index;
-                             queue_front_ptr[0]++;
-                         });
+        cgh.parallel_for(sycl::range<1>{ std::size_t(1) }, [=](sycl::id<1> idx) {
+            observation_indices_ptr[index] = value;
+        });
     });
 }
 
+///  A function that sets the queue sizes in ndview<std::int32_t, 1>& queue_size
```
Comment (Vika-F) [inline cpp/oneapi/dal/algo/dbscan/backend/gpu/kernel_fp_impl.hpp:703]: Need to distinguish somehow the SYCL queue and data point queue.
This description can be misleading.

**Review by Vika-F — COMMENTED** [review]
Thank you for adding the comments. It became much better now.
Still I have a couple of minor comments. Mostly related to cosmetic changes.

File: cpp/oneapi/dal/algo/dbscan/test/fixture.hpp [hunk 8]
```diff
@@ -46,12 +48,50 @@ class dbscan_test : public te::crtp_algo_fixture<TestType, Derived> {
...
+                if (aerr < tol || (!std::isfinite(lval) && !std::isfinite(rval)))
+                    continue;
+
+                const auto den = std::max({ eps, //
+                                            std::abs(lval),
+                                            std::abs(rval) });
+
+                const auto rerr = aerr / den;
+                CAPTURE(aerr, rerr, den, r, c, lval, rval);
+                REQUIRE(rerr < tol);
```
[the remark this answers is not in the document]
Comment (Alexandr-Solovev) [reply cpp/oneapi/dal/algo/dbscan/test/fixture.hpp:86]: Its done in the same style for all of our algorithms, in my understanding this check is necessary for doublechecking specific cases, when some of the values is ->0 or infinity

**Review by Vika-F — APPROVED** [review]
LGTM

**Pull request merged**